### PR TITLE
tests: added tests for staking

### LINKED ISSUE #487: Add more tests for Staking
Add more tests for:
- `src/dollar/Staking.sol`
- `src/dollar/StakingShare.sol`

Test coverage right now:
| File                                               | % Lines           | % Statements      | % Branches        | % Funcs          |
|----------------------------------------------------|-------------------|-------------------|-------------------|------------------|
| src/dollar/Staking.sol                             | 84.62% (110/130)  | 85.33% (128/150)  | 57.14% (16/28)    | 66.67% (16/24)   |
| src/dollar/StakingFormulas.sol                     | 100.00% (9/9)     | 100.00% (12/12)   | 100.00% (2/2)     | 100.00% (4/4)    |
| src/dollar/StakingShare.sol                        | 75.00% (30/40)    | 71.43% (30/42)    | 33.33% (2/6)      | 71.43% (10/14)   |

### PR DESCRIPTION
**Pull request by FibrinLab**
Resolves #487 

@rndquu Added some more tests. Most are not quite easy to run locally, I wonder why?

### CODE REVIEW

**Comment by rndquu** [comment]
> Resolves #487
> 
> @rndquu Added some more tests. Most are not quite easy to run locally, I wonder why?

Tests inherited from `LiveTestHelper` are supposed to run on a forked chain like `forge test --rpc-url https://cloudflare-eth.com `

Could you:
1. fix the failing tests here https://github.com/ubiquity/ubiquity-dollar/actions/runs/3971518930/jobs/6808501691
2. provide the updated test coverage for Staking and StakingShares so that we could compare the difference

**Comment by Steveantor** [comment]
Prefix any tests that require external RPC with `Remote`,
but make sure to only prefix those which fail locally.

see: https://github.com/ubiquity/ubiquity-dollar/blob/cdb91f66e98c973d16acab396ffe875b4b4ddc63/packages/contracts/test/dollar/Staking.t.sol#L69

**Comment by FibrinLab** [comment]
@rndquu Added more tests for staking and stakingShares. I am having issues with conventional commits though.

| File                          | %Lines                       | %Statements              | %Branches     | %Funcs
| ------------------ | ---------------------| ------------------- | ------------------ | ------------------ |
| Staking.sol             | 88.33% (106/120)     | 89.33% (134/150)   | 60.71% (17/28)    | 87.50% (21/24)   |
| StakingShares.sol | 77.50% (31/40)          | 73.81% (31/42)        | 33.33% (2/6)      | 78.57% (11/14)   |

**Comment by Steveantor** [comment]
> @rndquu Added more tests for staking and stakingShares. I am having issues with conventional commits though.
> 
> | File                          | %Lines                       | %Statements              | %Branches     | %Funcs
> | ------------------ | ---------------------| ------------------- | ------------------ | ------------------ |
> | Staking.sol             | 88.33% (106/120)     | 89.33% (134/150)   | 60.71% (17/28)    | 87.50% (21/24)   |
> | StakingShares.sol | 77.50% (31/40)          | 73.81% (31/42)        | 33.33% (2/6)      | 78.57% (11/14)   |
> 

Rebase `tests:...` to `test:...`
to pass the CC.

**Comment by FibrinLab** [comment]
> > @rndquu Added more tests for staking and stakingShares. I am having issues with conventional commits though.
> > 
> > 
> > 
> > File
> > %Lines
> > %Statements
> > %Branches
> > %Funcs
> > 
> > 
> > 
> > 
> > Staking.sol
> > 88.33% (106/120)
> > 89.33% (134/150)
> > 60.71% (17/28)
> > 87.50% (21/24)
> > 
> > 
> > StakingShares.sol
> > 77.50% (31/40)
> > 73.81% (31/42)
> > 33.33% (2/6)
> > 78.57% (11/14)
> 
> Rebase `tests:...` to `test:...` to pass the CC.

Done

**Comment by 0x4007** [comment]
I just asked GPT3 about fixing the verb in the commit message and it suggested:

```bash
git log --oneline | awk '{print $1}' | while read line; do
  if [[ $line != "fix:" && $line != "feat:" && $line != "chore:" && $line != "docs:" && $line != "style:" && $line != "refactor:" && $line != "test:" && $line != "build:" && $line != "ci:" ]]; then
    git commit --amend --no-edit --allow-empty
  fi
done
```

Update: it didn't work it did an infinite loop (seemed to work on another repo though, perhaps lint-staged messes with it.)

This shouldn't happen if your husky commit hook is setup. Make sure to install dependencies correctly.

**Comment by Steveantor** [comment]
> > > @rndquu Added more tests for staking and stakingShares. I am having issues with conventional commits though.
> > > File
> > > %Lines
> > > %Statements
> > > %Branches
> > > %Funcs
> > > Staking.sol
> > > 88.33% (106/120)
> > > 89.33% (134/150)
> > > 60.71% (17/28)
> > > 87.50% (21/24)
> > > StakingShares.sol
> > > 77.50% (31/40)
> > > 73.81% (31/42)
> > > 33.33% (2/6)
> > > 78.57% (11/14)
> > 
> > 
> > Rebase `tests:...` to `test:...` to pass the CC.
> 
> Done

do it for all

**Comment by FibrinLab** [comment]
> > > > @rndquu Added more tests for staking and stakingShares. I am having issues with conventional commits though.
> > > > File
> > > > %Lines
> > > > %Statements
> > > > %Branches
> > > > %Funcs
> > > > Staking.sol
> > > > 88.33% (106/120)
> > > > 89.33% (134/150)
> > > > 60.71% (17/28)
> > > > 87.50% (21/24)
> > > > StakingShares.sol
> > > > 77.50% (31/40)
> > > > 73.81% (31/42)
> > > > 33.33% (2/6)
> > > > 78.57% (11/14)
> > > 
> > > 
> > > Rebase `tests:...` to `test:...` to pass the CC.
> > 
> > 
> > Done
> 
> do it for all

I fixed it now

**Comment by 0x4007** [comment]
Thanks @FibrinLab. Small request but can you bump up the coverage to 80% minimum? We defined 80% minimum test coverage on the other issue but it seems like we have forgotten to add it to this issue. I think it could be pretty quick since you're already at 77% for `StakingShares.sol` according to the table you included in your comment.

@rndquu for any of the existing "upgrade tests" related issues please specify 80% minimum and continue to do so for any future bounties created for upgrading tests.

File: packages/contracts/test/dollar/StakingShare.t.sol [hunk 1]
```diff
@@ -48,6 +51,12 @@ contract RemoteDepositStateTest is DepositState {
     uint256[] ids;
     uint256[] amounts;
 
+    // function testCannotDeployEmptyManager() public {
```
Comment (rndquu) [inline packages/contracts/test/dollar/StakingShare.t.sol]: I think it is better to remove the commented code

File: packages/contracts/test/dollar/StakingShare.t.sol [hunk 2]
```diff
@@ -90,6 +170,36 @@ contract RemoteDepositStateTest is DepositState {
         assertEq(stakingShare.holderTokens(secondAccount), ids);
     }
 
+    function testCannotSafeTransferFromUnapproved() public {
```
Comment (rndquu) [inline packages/contracts/test/dollar/StakingShare.t.sol]: This is already tested in the parent `ERC1155` contract. I think the only thing that should be tested here is that the holder balance was updated https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/StakingShare.sol#L159

File: packages/contracts/test/dollar/StakingShare.t.sol [hunk 3]
```diff
@@ -112,16 +222,40 @@ contract RemoteDepositStateTest is DepositState {
         assertEq(stakingShare.holderTokens(secondAccount), ids);
     }
 
+    function testCannotBatchTransferFromUnapproved() public {
```
Comment (rndquu) [inline packages/contracts/test/dollar/StakingShare.t.sol]: Again this is already tested by the parent `ERC1155` contract. Just test the holder balance https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/StakingShare.sol#L173

File: packages/contracts/test/helpers/LiveTestHelper.sol [hunk 4]
```diff
@@ -288,6 +288,7 @@ contract LiveTestHelper is Test {
 
         vm.startPrank(fourthAccount);
         dollarToken.approve(address(metapool), 10000e18);
+        // dollarToken.approve(address(fourthAccount), 1000e18);
```
Comment (rndquu) [inline packages/contracts/test/helpers/LiveTestHelper.sol]: Why is it commented? If we don't need this approve and tests are passing then pls just delete the line

File: packages/contracts/test/helpers/LiveTestHelper.sol [hunk 5]
```diff
@@ -333,6 +334,8 @@ contract LiveTestHelper is Test {
             locked
         );
 
+        deal(address(dollarToken), address(staking), 100e18);
```
Comment (rndquu) [inline packages/contracts/test/helpers/LiveTestHelper.sol:336]: This is kind of a hack because it sets the staking contract's balance out of nothing. I mean in a normal flow user should deposit tokens for the staking balance to increase. I think it is clearer to increase staking contract's balance vie the `deposit` method. When a new dev opens the tests he wants to see the normal flow to understand the entire staking mechanic.

**Review by rndquu — CHANGES_REQUESTED** [review]
@FibrinLab 

Thanks for the PR

Overall we need more tests. Before the PR the avg test coverage for staking and staking shares were `68.11%`, now it is `73.63%`

We expect at least `80%`

File: packages/contracts/test/dollar/StakingShare.t.sol (hunk 1 above)
Comment (FibrinLab) [reply packages/contracts/test/dollar/StakingShare.t.sol]: Done

File: packages/contracts/test/dollar/StakingShare.t.sol (hunk 2 above)
Comment (FibrinLab) [reply packages/contracts/test/dollar/StakingShare.t.sol]: Removed

File: packages/contracts/test/helpers/LiveTestHelper.sol (hunk 4 above)
Comment (FibrinLab) [reply packages/contracts/test/helpers/LiveTestHelper.sol]: Removed

File: packages/contracts/test/dollar/StakingShare.t.sol (hunk 3 above)
Comment (FibrinLab) [reply packages/contracts/test/dollar/StakingShare.t.sol]: Removed

File: packages/contracts/test/helpers/LiveTestHelper.sol (hunk 5 above)
Comment (FibrinLab) [reply packages/contracts/test/helpers/LiveTestHelper.sol:336]: okay. got it

**Comment by FibrinLab** [comment]
@rndquu @pavlovcik Please have a look at this. I added more functions for both the staking and stakingShares contracts. However, I noticed 2 functions (_burn and _burnBatch), both internal functions from the staking contracts. These 2 functions are not called or used by any other external functions in the stakingShares contract. Commenting them together with the updates I made gives this new coverage report. Please I need recommendations.

| File                          | %Lines                       | %Statements              | %Branches     | %Funcs
| ------------------ | ---------------------| ------------------- | ------------------ | ------------------ |
| Staking.sol             | 88.33% (106/120)     | 89.33% (134/150)   | 60.71% (17/28)    | 87.50% (21/24)   |
| StakingShares.sol | 96.88% (31/32)          | 96.88% (31/32)       | 100% (2/2)          | 91.67% (11/12)   |

**Comment by 0x4007** [comment]
I'm not the one to say with confidence but superficially the numbers you provided look good. @zgorizzo69 @0xcodercrane @rndquu @Steveantor any remarks?

One last remark, but I'm currently working on refactoring the UI code and ended up down in the staking test file. 

I don't think we need to be testing for "v1" of anything because those have been deprecated. Focus only on "v2" although on my branch I'm removing the versioning and just calling it "Staking" and "StakingToken"

**Comment by rndquu** [comment]
> @rndquu @pavlovcik Please have a look at this. I added more functions for both the staking and stakingShares contracts. However, I noticed 2 functions (_burn and _burnBatch), both internal functions from the staking contracts. These 2 functions are not called or used by any other external functions in the stakingShares contract. Commenting them together with the updates I made gives this new coverage report. Please I need recommendations.
> 
> File	%Lines	%Statements	%Branches	%Funcs
> Staking.sol	88.33% (106/120)	89.33% (134/150)	60.71% (17/28)	87.50% (21/24)
> StakingShares.sol	96.88% (31/32)	96.88% (31/32)	100% (2/2)	91.67% (11/12)

Yes, it seems that `_burn` and `_burnBatch` are not used anywhere so you can delete them

Overall the PR looks good in terms of minimal coverage

I guess in the next test iteration milestone (not current issue) we could add some intergration tests for staking and cover all edge cases for staking to make coverage 100%

File: packages/contracts/test/helpers/LiveTestHelper.sol (hunk 5 above)
Comment (0x4007) [reply packages/contracts/test/helpers/LiveTestHelper.sol:336]: This is the only unresolved conversation. Looks like the fix was not implemented yet?

Comment (FibrinLab) [reply packages/contracts/test/helpers/LiveTestHelper.sol:336]: its fixed

**Comment by FibrinLab** [comment]
> > @rndquu @pavlovcik Please have a look at this. I added more functions for both the staking and stakingShares contracts. However, I noticed 2 functions (_burn and _burnBatch), both internal functions from the staking contracts. These 2 functions are not called or used by any other external functions in the stakingShares contract. Commenting them together with the updates I made gives this new coverage report. Please I need recommendations.
> > File	%Lines	%Statements	%Branches	%Funcs
> > Staking.sol	88.33% (106/120)	89.33% (134/150)	60.71% (17/28)	87.50% (21/24)
> > StakingShares.sol	96.88% (31/32)	96.88% (31/32)	100% (2/2)	91.67% (11/12)
> 
> Yes, it seems that `_burn` and `_burnBatch` are not used anywhere so you can delete them
> 
> Overall the PR looks good in terms of minimal coverage
> 
> I guess in the next test iteration milestone (not current issue) we could add some intergration tests for staking and cover all edge cases for staking to make coverage 100%

Alright. I deleted it. Thanks for the review

**Review by 0x4007 — APPROVED** [review]

**Review by rndquu — APPROVED** [review]
